### PR TITLE
[metadata] make get_darwin_locale thread-safe

### PR DESCRIPTION
**Pull request by lewurm**
fixes https://bugzilla.xamarin.com/show_bug.cgi?id=59916

### CODE REVIEW

File: mono/metadata/locales.c [hunk 1]
```diff
@@ -515,7 +516,8 @@ get_darwin_locale (void)
 		CFRelease (locale);
 	}
 
-	return g_strdup (darwin_locale);
+	mono_memory_barrier ();
+	return g_strdup (cached_locale = darwin_locale);
```
Comment (luhenry) [inline mono/metadata/locales.c]: You would need to us `mono_atomic_cas_ptr` to make sure you do not leak memory here.

Comment (kumpera) [reply mono/metadata/locales.c]: This is a convergent leak, it's fine.

Comment (vargaz) [inline mono/metadata/locales.c]: This looks a bit confusing, should be split to 2 lines.

Comment (lewurm) [reply mono/metadata/locales.c]: I agree with @kumpera. on my system it's six bytes and anyways this case is pretty unlikely.

File: mono/metadata/locales.c [hunk 2]
```diff
@@ -515,7 +516,9 @@ get_darwin_locale (void)
 		CFRelease (locale);
 	}
 
-	return g_strdup (darwin_locale);
+	mono_memory_barrier ();
+	cached_locale = darwin_locale;
+	return g_strdup (cached_locale);
```
Comment (marek-safar) [inline mono/metadata/locales.c]: As the assignment is not atomic, this should be `g_strdup (darwin_locale);`

**Comment by marek-safar** [comment]
Please backport to 2017-10

File: mono/metadata/locales.c (hunk 2 above)
Comment (vargaz) [reply mono/metadata/locales.c]: Pointer assignments are atomic.

**Pull request merged**